domain/application/service/charm.go

```diff
@@ -530,45 +547,242 @@ func (s *Service) SetCharmAvailable(ctx context.Context, id corecharm.ID) error
+}
+
+// GetAvailableCharmArchiveSHA256 returns the SHA256 hash of the charm archive
+// for the given charm id. If the charm is not available,
+// [applicationerrors.CharmNotResolved] is returned.
+func (s *Service) GetAvailableCharmArchiveSHA256(ctx context.Context, id corecharm.ID) (string, error) {
+	if err := id.Validate(); err != nil {
+		return "", fmt.Errorf("charm id: %w", err)
+	}
+	return s.st.GetAvailableCharmArchiveSHA256(ctx, id)
+}
+
+// ResolveUploadCharm resolves the upload of a charm archive. If the charm is
+// being imported from a migration then it can returns
+// [applicationerrors.CharmNotFound]. Returns
+// [applicationerrors.CharmSourceNotValid] if the source is not valid. Returns
+// [applicationerrors.CharmMetadataNotValid] if the charm metadata is not valid.
+// Returns [applicationerrors.CharmNameNotValid] if the reference name is not
+// valid. Returns [applicationerrors.CharmManifestNotFound] if the charm
+// manifest is not found. Returns [applicationerrors.CharmDownloadInfoNotFound]
+// if the download info is not found. Returns [applicationerrors.CharmNotFound]
+// if the charm is not found.
+func (s *Service) ResolveUploadCharm(ctx context.Context, args charm.ResolveUploadCharm) (charm.CharmLocator, error) {
+	var localUploadCharm bool
+
+	switch args.Source {
+	case corecharm.CharmHub:
+		if !args.Importing {
+			return charm.CharmLocator{}, applicationerrors.NonLocalCharmImporting
+		}
+	case corecharm.Local:
+		localUploadCharm = !args.Importing
+	default:
+		return charm.CharmLocator{}, applicationerrors.CharmSourceNotValid
+	}
+
+	// We're not importing a charm, this is a full blown upload.
+	if localUploadCharm {
+		return s.resolveLocalUploadedCharm(ctx, args)
+	}
+
+	// We're importing a charm, all the metadata of the charm will have been
+	// migrated already, so this is pushing the charm to the object store and
+	// setting the charm as available.
+	return s.resolveMigratingUploadedCharm(ctx, args)
+}
+
+func (s *Service) resolveLocalUploadedCharm(ctx context.Context, args charm.ResolveUploadCharm) (charm.CharmLocator, error) {
+	result, digest, err := s.charmStore.StoreFromReader(ctx, args.Reader, args.SHA256Prefix)
+	if err != nil {
+		return charm.CharmLocator{}, internalerrors.Errorf("resolving uploaded charm: %w", err)
+	}
+
+	// We must ensure that the objectstore UUID is valid.
+	if err := result.ObjectStoreUUID.Validate(); err != nil {
+		return charm.CharmLocator{}, internalerrors.Errorf("invalid object store UUID: %w", err)
+	}
+
+	// Note: we don't have a full SHA256 from the user so, we can't actually
+	// check the integrity of the charm archive. We can only check the prefix
+	// of the SHA256 hash (that's enough to prevent collisions).
+
+	// Make sure it's actually a valid charm.
+	ch, err := internalcharm.ReadCharmArchive(result.ArchivePath)
+	if err != nil {
+		return charm.CharmLocator{}, internalerrors.Errorf("reading charm archive %q: %w", result.ArchivePath, err)
+	}
```

Comment:
This should be changed. This is the only place we read `ArchivePath`, which turns out to be the path we store the charm in a tmp dir before passing it on to the object store.

We're trying to move away from `ArchivePath` entirely, treating it as just a natural key for the object store metadata service atm, so I think this is a misuse.

Removing these will mean we can drop `ArchivePath` from the `result` entirely.

We could do this by `Tee`ing out the args.Reader when it's read to be stored by the `charmStore`.

There is the danger of exerting too much memory pressure, since we may end up caching the contents of reader (i.,e. the charm blob) in a buffer. If we're really clever, we may be able to find a way to do both at the same time?